### ISSUE TITLE
[xml] Please externalize the mapping of file extensions to languages as preference file 

### ISSUE THREAD

**Issue by tkleiber** [pmd/pmd#461]
This is reopening of:
https://sourceforge.net/p/pmd/feature-requests/568/

**Rule Set:**
all

**Description:**
Especially for xml are a lot of file extensions possible, which all represents some kind of "DSL"'s:

 - configuration files of several ide's and frameworks, like
  - Oracle JDeveloper (jws, jpr)
  - Oracle ADF (cpx, xcfg, dcx, jpx)
 - xsd
 - wsdl
 - several specialized XML message standards, eg. FpML (Financial products Markup Language)
 - etc.

For this is would useful if the mapping is configurable externally in a file, which can be written from a text editor or from an ide plugin.

Maybe this is relevant for other languages too.

**Code Sample demonstrating the issue:**
Actually the code is static in XmlLanguageModule:
        super(NAME, null, TERSE_NAME, XmlRuleChainVisitor.class, "xml");

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

**Comment by jsotuyod** [comment]
This one is an interesting request... Several generic rules should apply to all, but I can think of having special rules just for one or the other; but always using the same parser / AST.

Kind of having ad-hoc definitions of "language versions", with different extensions.

This needs some thinking and refactoring, but seems interesting on it's own.

**Comment by tkleiber** [comment]
Find additional use case: jsp contains rulesets for jsf, but allows only extensions for jsp, not for jsf.

**Comment by jordillachmrf** [comment]
I think that support for .tag and .jspf files also fall into this issue. Any plan to implement this ?

**Comment by brimstony** [comment]
I just started digging into PMD and am new to the party, but this is a deal breaker for me as well.

**Comment by hooperbloob** [comment]
Suggest we retain the internal mapping but provide for an optional external one that can override it.

**Comment by oowekyala** [comment]
I think 7.0.0 is a good target release for this feature, especially because of #1540

The main thing to get straight before getting started is the details of *how* to configure it. Open questions:
* Should it be a CLI/Maven/etc parameter? I think yes, and that's what this issue is about 
  * But what should the format be? I think specifying them inline on the CLI wouldn't scale, so probably we should use an external config file. But in what format? 
    * XML seems the easiest to support: verbose, but can be parsed readily without external library. Would also be easily embeddable into the ruleset's XML, see next points
    * Other candidates: YAML, TOML, `.properties`. In any case we should be able to change the mapping for each language individually, which means that we need some kind structure. A flat .gitignore-like file wouldn't work
* Should it be configurable in the ruleset? This would make the mapping portable
* Should it be configurable on a per-rule basis? This make sense especially for XML rules, because different XML file types (ie schema) mandate and exclude different sets of rules
  * If so, how to make just a few rules in a bigger ruleset share the same custom mapping, instead of duplicating it? 
    * Allow giving names to mappings declared ruleset-wide and reference them from rules? 
    * Split those rules into another ruleset with the custom mapping? How would the include mechanism work? (see next point)
* How should supplying several mappings (eg one on CLI, one local to the ruleset, one local to a rule in the ruleset) behave? Should e.g. the ruleset mapping override the CLI-supplied mapping, or the reverse?
  * What if PMD is run with several rulesets? 
  * What if a ruleset includes a ruleset that has its own mapping? 
  * Should custom mappings override completely the default mapping? Or just augment it with some new mappings without removing the default ones? Both have use cases...
* Implementation-wise: 
  * Should providing new mappings override the mappings in the language object (which is static to the VM)? Changing this static state will probably mess with PMD's programmatic API, used e.g. in IDE plugins 
  * Should it create a new Language with those mappings? Meaning that other language uses the parser and AST of the base language, but has its own custom file mapping.
    * This would allow us to avoid creating new language modules when the AST is equivalent (looking at those `pom`, `xml`, `xsl`, etc modules under `pmd-xml`), and make instead the language registry extensible.
    * This would go hand in hand with the suggestion above "Allow giving names to mappings declared ruleset-wide and reference them from rules"
    * Should those be registered on the LanguageRegistry, or only local to one PMD run? Again, registering those statically is probably a bad idea, but the alternative probably would probably need big code changes

**Comment by adangel** [comment]
Currently, these file extensions are only relevant when executing PMD from commandline - then PMD itself collects the files to be analyzed. In maven-pmd-plugin, PMD is provided with a ready list of files (DataSources). For ant, you specify a fileset.
One point to think about is: Do we need the file extensions at all? If possible, we should try to avoid the need to configure the file extensions per language at all. The less option, the more simpler it gets. However, we might need it, if we want to support a ruleset which contains rules for multiple languages and you want to analyze the files in one PMD run (with multiple runs, you could use multiple rulesets).
Each rule is currently only applied, if the filename is not in the excluded list defined by the ruleset. So, if you currently use a ruleset, which contains rules for more than one language, you already have the problem, that each rule is executed for all files... (which means, we do not currently support mixing languages in a single ruleset).

The question is maybe: How does PMD select the files to be processed?

**Comment by kris-scheibe** [comment]
I think a more basic question is _if_ PMD should try to match files/filetypes and rules.

By removing the extension filtering, PMD would just apply every rule to every given file. This would presumably also work for a majority of PMD users who run a single type of rules on a single type of files anyway.
(Also this would break setups where a single PMD run processes different files with different rulesets -- even though this might not be that big of a concern for new major release.)

Despite that I would rather give PMD a more powerful (and also user-configurable) mechanism to match files, languages and rules:
Using the file extension works well with code files and poorly when it comes to container formats (XML, JSON or YAML). MIME for example, uses a suffix such as `application/wsdl+xml` or `application/hal+json` to differentiate the format and the content of a file.

One possible approach is to have a ruleset both declare the language to use **and** the files/extensions it applies to independently:
```xml
<ruleset name="My custom DSL" filepattern="**/custom-dsl/*.something" ...>   
	<rule ref="category/xml/errorprone.xml/MistypedCDATASection" />

	<rule name="SomeCustomDSLRule" language="xml" ...>
	...
	</rule>
</ruleset>
```
(Or maybe use/extend `<include-pattern>` and `<exclude-pattern>`?)

The current extensions could act as a fallback so that rules that only declare a language still work.


Given this kind of approach, I don't think there should be a separate CLI/Maven/etc parameter independent of the rule/ruleset either: 

To define a different file extension for existing rules instead of running PMD with
`-R category/jsp/design.xml/NoInlineScript,category/jsp/design.xml/NoInlineStyleInformation`
or
```xml
<rulesets>
	<ruleset>category/jsp/design.xml/NoInlineScript</ruleset>
	<ruleset>category/jsp/design.xml/NoInlineStyleInformation</ruleset>
</rulesets>
```

You could create a custom ruleset to define the extension:
```xml
<ruleset name="Java Server Everything" filepattern="*.jsp;*.jspf;*.jsf;*.tag" ...>
	<rule ref="category/jsp/design.xml/NoInlineScript"/>
	<rule ref="category/jsp/design.xml/NoInlineStyleInformation"/>
</ruleset>
```
and run PMD with that ruleset instead.


When dealing with included rules, rulesets and sub-rulesets I would enforce having one language and one file mapping per top-level ruleset.
So either all included rules and rulesets define the same language and file mapping, which the top-level ruleset would implicitly use, or the top-level ruleset defines file mapping and overrides the values in the included rulesets.
The override mechanism is also needed if an included ruleset already defines a language and filepattern (e.g. `language="xml"` and `filepattern="*.wsdl"`) and those rules should check different files (neither *.xml or *.wsdl).


So these are my 2 Cents (maybe a bit more) on how to possibly add this feature without throwing out too much existing code and keeping existing rules somewhat compatible...

**Comment by oowekyala** [comment]
> One possible approach is to have a ruleset both declare the language to use and the files/extensions it applies to independently
> When dealing with included rules, rulesets and sub-rulesets I would enforce having one language and one file mapping per top-level ruleset.

I like your approach. I would also enforce that all rules in a ruleset apply to a single language, declared explicitly by the ruleset. As @adangel said, we do not currently support mixing languages in a single ruleset, and enforcing that that doesn't happen gives us a strong organisation principle for free. So basically:
* one ruleset may only contain rules for a single language
  * included rulesets must have the same language, so this is enforced recursively
* running pmd for several languages means creating a ruleset for each of them
  * This doesn't mean each ruleset has to be run in a separate PMD execution, right? We can already execute several rulesets at once (they're merged in a Rulesets). The difference here would be that we don't merge anything. Each ruleset included on the top level (e.g. with the `-R` option) has its own set of file filters. We dispatch each file to all rulesets that cover it.

About XML syntax, I'd be for merging the `<include-pattern>` and `<exclude-pattern>` we currently have on the top-level with the default language extensions, to have a single mechanism to filter files for a ruleset. I'd also wrap them all in an element for clarity. We could for example add a header element to the ruleset declaration, which specifies:
* The language the ruleset applies to
* The include and exclude patterns. The following example uses a flat .gitignore-like style to cut on boilerplate.

```xml
<ruleset name="Java Server Everything" ...>
  <domain language="jsp"> <!-- Rules referenced and defined in this ruleset are checked to be JSP rules -->
     <include-patterns> <!-- These default to the file extensions we have at the moment -->
**/*.jsp
**/*.jspf
**/*.jsf
**/*.tag
     </include-patterns>
     <exclude-patterns> <!-- These default to none -->
...
     </exclude-patterns>
  </domain>
```

About the inclusion/override mechanism:

> So either all included rules and rulesets define the same language and file mapping, which the top-level ruleset would implicitly use, or the top-level ruleset defines file mapping and overrides the values in the included rulesets.

Probably a simpler scheme, which also makes everything more explicit, is not to use the file filters of included rulesets at all (i.e. a ruleset always uses only its own `domain`). This prevents confusion when several rulesets are included, or when a single rule is included from another ruleset which has custom file filters. We could add a "keep-filters" option on the ruleset inclusion XML tag later-on if there's a use case, but IMO this seems general enough. Wdyt? 

> Do we need the file extensions at all? If possible, we should try to avoid the need to configure the file extensions per language at all.
 
With this scheme, the file filtering is configured by ruleset, not by language, and Language objects only provide defaults.

**Comment by kris-scheibe** [comment]
I think it would be better to not initially enforce the declaration of the language/patterns of a ruleset and just warn, if a ruleset doesn't explicitly define it:
* For any ruleset that contains actual rule-definitions (and not just references to other rules or rulesets) we can derive the language (and it's file pattern).
* A ruleset that contains different languages should fail to load, turning the current (undocumented) behavior into a proper error.
* We could even include the expected declaration in the warning message to allow users to migrate their rulesets without having to dig through the documentation.
```
[WARNING] RuleSet domain is missing. Future versions of PMD will require it.
RuleSet /path/to/ruleset.xml will use this implicit declaration:
<domain lang="java">
   ...
</domain>
```
The users will have a much easier time migrating their existing rulesets to PMD 7 if we don't require them to add the domain declaration, especially if the rules are fully compatible apart from that.

**Regarding the inclusion/override mechanism:**
I'm not sure we should have only place to define inclusion patterns for both paths and file extensions:
* Assuming the default PMD rulesets each have a properly defined domain and patterns, e.g. `<include-patterns>**/*.jsp</include-patterns>`.
* If a user wants to add new extensions, they could simply override the definition as seen in the `Java Server Everything`-example above.
* If they only want to set a directory that should be included, while keeping the extension, they have to define the extension _and_ the directory. Defining `some/directory/**` and `**/*.jsp` separately (or by a "keep-filters" mechanism) would include all `jsp`-files and everything in `some/directory`, so the definition would have to be `some/directory/**/*.jsp`.
* A ruleset, that has multiple extensions and multiple directories, can become quite complex.

Even though the pattern to match a filename and the pattern to match the directory the file is in, are technically the same, they represent different use-cases:
The filename (or extension) defines the type of files that should be processed. For code-files, the allowed extensions are usually strictly defined. Template-languages have a broader range of commonly used extensions (i.e jsp/jspf/jsf/tag/...). For languages that act as a container for freely definable content (i.e. XML/JSON/YAML) the filename depends mostly on the DSL and has few, if any, restrictions otherwise.
The directory is used to have a ruleset process or not process a certain set of files so it's more used as a high-level grouping.
Or as a TLDR: The filename makes Java-rules process only ".java" files, the directory lets you run a ruleset on one package only or on everything except one package.


If we split the includes into `<include-files>` and `<include-directories>` _and_ allow using the `<include-files>` of included rulesets, we can have the included ruleset define the default file-patterns which don't get overridden accidentally or have to be redefined. (Maybe it should even be an error, if an included ruleset defines `<include-directories>` or any excludes.)

So we could have a default ruleset, like category/java/errorprone.xml:
```
<ruleset name="Error Prone" ...>
    <domain language="java">
        <include-files>
            *.java
        </include-files>
    </domain>
    ...
</ruleset>
```

And still redefine the directories (i.e. java-packages):
```
<ruleset name="My Java Rules" ...>
    <domain language="java">
        <include-directories>
            **/some/javapackage/**
            **/another/javapackage/**
        </include-directories>
    </domain>

    <rule ref="category/java/errorprone.xml"/>
</ruleset>
```

So inclusions/overrides would work something like this:
* A ruleset that doesn't contain a domain declaration and only references other rules/rulesets acts as a plain collection of other rules and doesn't influence the inclusions and overrides.
* A ruleset that has a domain (declared or implicitly derived from its rules) overrides the domain of all included rules and rulesets
    * A ruleset that declares a domain must also define the file patterns if it defines own rules and not just references other rules/rulesets.
    * A ruleset that declares directory includes/excludes may only be included by a collection-ruleset (first bullet-point). Directory includes/excludes can't override each other.
* A rule/ruleset may be referenced in different rulesets with different file patterns, includes and excludes. The files that the ruleset will process only depend on the parent-rulesets and definitions, each definition of one ruleset is independent of other definitions of the same ruleset.

In the end this would leave us with three "types" or "levels" of rulesets:
* "collection" rulesets: These only reference other rulesets and could be replaced by running PMD with the list of included rules/rulesets directly.
* "normal" rulesets: These define the language, rules, include/exclude directories, and/or file-extensions.
* "include" rulesets: These are the low-level collection of rules and should define the file-extensions, these rules usually run on.
    * "include" rulesets can also reference other "include" rulesets.


Implementation-wise it might be useful to not use the ruleset to group all rules that should be run on a specific set of files.
With this kind of definitions we would end up with one or multiple domains that describe a set of files to process and a list of rules that should be used.
If the same rule is used in different includes we'd even have a many-to-many relationship between domains and rules.

**Comment by oowekyala** [comment]
> ending up with a Set<Rule> (not just conceptually) seems like the logical consequence to me.

Just for extensibility reasons it's better to hide the fact we're using a Set behind our own interface. We could also be internally using a map or something else to precompute which rules apply to which file, something like that. For now the implementation with a List is very inefficient, mostly because we repeat many computations -and reusing the results means having auxilliary data structures that are better off hidden inside the ruleset object.

We could however potentially remove one layer of the cake by eliminating the Rulesets class (which aggregates rulesets), and instead merge every rule into a single ruleset to be run later on... Idk for sure  

> And un-including would work similarly by specifying <include-files /> on the rule-def?

Exactly

> So from="language" would revert to the language-defaults from="ruleset" would ignore the ruleset-level file-filters and retain the file-patterns of the included rule?

What I meant was, that `from="ruleset"` would use the file filters of the current ruleset and ignore those of the included rule. Think of it like this: a rule-ref's defaulting sequence is 
  1. itself
  2. its referenced rule (only for file patterns)
  3. the current ruleset 
  4. the language defaults 
  
If the ref doesn't mention file patterns, they're defaulted to the referenced rule implicitly (step 2). If it does mention file patterns, then we choose those (step 1) and nothing is defaulted. That means it can't ever reach step 3 or 4. That's why we need an explicit mechanism to jump over the implicit step. Having a `from="imported"` would be redundant because it would be equivalent to mentioning nothing (step 2 is implicit).

On the other hand for rule defs there's no step 2, but if the current ruleset (step 3) declares file patterns for the relevant language then they're used implicitly, so step 4 cannot be reached anymore, hence why `from="language"` could be useful. But `from="ruleset"` is implicit for ruledefs so redundant.

This is a bit like receiver shadowing in e.g. java or kotlin, where sometimes you have to qualify a `this` to refer to the outer instance. 

> Also a <file-patterns from="language" /> could be useful on the ruleset-level to prevent inheriting the file-pattern for all referenced rules.

This would be equivalent to mentioning nothing. If the ruleset doesn't mention a file-patterns they're already defaulted to the language defaults

> I don't think we need a different element, though. It's just a different way to define a rule-level file-pattern

From the point of view of schema design, it's probably more consistent to have two elements. In XSD 1.0 it's impossible to forbid adding content to an element based on the presence or not of an attribute. So schema validation (and so, your IDE) would not prevent you from writing e.g.
```xml
<file-patterns from="language">
pom.xml
</file-patterns>  
```
In that case, the interpretation would be left to PMD's ruleset parser. Should it throw an error? Should it append those new patterns to those you take from the language? There's a use case for that, but it could also be expressed by e.g.
```xml
<default-file-patterns from="language"/>
<file-patterns>
pom.xml
</file-patterns>  
```
I'd rather have very strong schema assertions and orthogonality between elements -but that's just a personal preference. Maybe it would be indeed simpler to not have too many different elements, idk. Those are my 2 cents on that anyway

> A completely identical definition and reference-path (multiple verbatim definitions in one file). -> Warn when loading the ruleset, otherwise drop the duplicates

> An identical definition but through different reference chains (A -> B -> X and A -> C -> X) ... -> warn in debug mode (I wouldn't nag the user to import every rule exactly once), otherwise drop the duplicates

I completely agree with that

> Rule 3 would check pom.xml in folder A twice

The simplest solution that comes to mind would be to keep track of which rule scanned what file in each directory, clearing the index every time we change directory. Although I'm not certain how PMD dispatches files, especially when running with multiple threads, so I don't know how realistic that is...

**Comment by jsotuyod** [comment]
> Each rule is currently only applied, if the filename is not in the excluded list defined by the ruleset. So, if you currently use a ruleset, which contains rules for more than one language, you already have the problem, that each rule is executed for all files... (which means, we do not currently support mixing languages in a single ruleset).
> 
> The question is maybe: How does PMD select the files to be processed?

I'm unsure I follow... I personally DO have a single ruleset with JSP, XML, JS and Java rules in place for a single project and PMD applies the rules that correspond to each file according to it's extension. This is a behavior I actually rely on. I'm doing this on a Maven project, as by default it setups a single run (Gradle is way easier to work with).

1. `RuleSets.applies` just checks if any `RuleSet.applies` is true.
2. `RuleSet.applies` just checks include / excludes (which by default are both empty and allows everything)
3. upon calling `apply`, the `RuleSet` actually checks if **each rule** applies to the language version inferred based on the file extension

https://github.com/pmd/pmd/blob/e57d365da905e619500f75b32e4735cfc8c21c85/pmd-core/src/main/java/net/sourceforge/pmd/RuleSet.java#L530-L536

I see lots of the conversation has geared on top of the assumption this behavior is broken, but it isn't.


I acknowledge that XML in particular is troublesome due to the number of files using XML structure, but I would refrain from forcing upon the user to specify this for all rulesets. That is a huge regression in usability and moves the burden back to the user. **I'm pro for extensibility.**

I don't personally think we have a problem with the current mechanics, but I *do* think we have a problem with the defaults. Actually, our inferring acknowledges that a file may correspond to more than one language:

https://github.com/pmd/pmd/blob/e57d365da905e619500f75b32e4735cfc8c21c85/pmd-core/src/main/java/net/sourceforge/pmd/lang/LanguageVersionDiscoverer.java#L102-L105

A `pom.xml` may be a POM file, but is also an XML, and all XML files should apply.

Where we screw up is right here, when we try to turn from language to language version:

https://github.com/pmd/pmd/blob/e57d365da905e619500f75b32e4735cfc8c21c85/pmd-core/src/main/java/net/sourceforge/pmd/lang/LanguageVersionDiscoverer.java#L79

We try to reduce it back again to a single option, which of course may not be true.

**Comment by tkleiber** [comment]
No unfortunatly #3417  does not solve my problem,. I have a monorepo with mixed technology. So I want to run pmd over this whole codebase one time. 

Espcially for the mentioned xml file extensions above we have a lot of custom XPath rules. Normally I change the mentioned code line to match our file extensions as xml and rebuild pmd therefore. And I have not does this some time now, so we stuck with an older version of pmd at the moment, so I don't know if this is the current code anymore. 

Have seen the same problem with PL/SQL , but not use it till now.
